zerver/migrations/0482_automatically_follow_unmute_topics_policy_defaults.py

```diff
@@ -0,0 +1,90 @@
+# Generated by Django 4.2.6 on 2023-10-20 23:43
+
+from django.db import migrations, models
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+from django.db.models import Max
+
+AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_SEND = 2
+AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_INITIATION = 3
+
+
+def set_default_user_topic_policies(
+    apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
+) -> None:
+    RealmUserDefault = apps.get_model("zerver", "RealmUserDefault")
+    UserProfile = apps.get_model("zerver", "UserProfile")
+
+    # Since we have 'atomic=False', we use '+ 1000' to safely backfill
```

Comment:
The "safely" in this comment implies that this algorithm is safe.  It is _still not safe_ -- since defaults are not applied at the database level, but rather in code, any users created after this migration is applied but before Django is restarted into the new code will still be created with the old defaults if they had the old version of 0467/0477.  This isn't a serious problem with this migration, but I am a little wary of a comment that implies that this sort of solution is correct in general.  

This comment is also misleading, since being `atomic=True` **would not prevent the possibility**.  PostgreSQL's [default isolation level is "read committed", which does not prohibit phantom reads](https://www.postgresql.org/docs/current/transaction-iso.html), where two queries in a transaction can return different sets of rows depending what transactions have committed in the meantime.

Reply:
My understanding is that, the backfill operation (which the function does) is always safe?

For zulip cloud:
For any user created after 0476 and before 0482, the backfill function will update the NULL with the new value.

For servers already having the old version of 0476 and 0477:
Any new user created will have the old default, so the backfill operation can safely update the value in this case too.

As you already mentioned, the only case where the the user creation will fail on zulip cloud is after 0482 and before django is restarted.

Am I missing something?

Reply:
"Is not safe" may have been misleading here -- I meant "is not guaranteed to update every user."  This is a hypothetical, because Zulip Cloud does not currently have 0476/0477 applied -- but let's assume that production was currently running 0477.

1. Staging starts its deploy.  Production continues serving production traffic while this happens.
2. Staging applies database changes, including 0482.  This updates every row currently in the database.
3. It also adjusts the Django-level "default," but this migration step _is effectively a no-op_ -- since the only thing that establishes a default at creation time is the code in the running Django process, and production is still running old code.
4. A new user is created in production.  The Django-level default in its running code is still `NEVER`, so this new user gets that default value.
5. Production does its deploy.  The django processes are restarted into the new code, picking up the new Django-level default.
6. A new user is created in production.  The Django-level default is no longer `NEVER`, so this user gets that value.

We're left with some number of users (between steps 3 and 5, which may be minutes to hours) which did not pick up the default we wanted from the backfill.

Again, since Zulip Cloud does not have 0467/0477 deployed, this isn't a practical problem with this migration.  But it's a problem with this _sort_ of migration, and I don't want misleading comments left in the codebase which imply to anybody that tries a similar thing later that this will _always_ be safe.